backend/student_on_tour/views.py

```diff
@@ -159,3 +164,71 @@ def get(self, request):
         return Response(serializer.data, status=status.HTTP_200_OK)
+
+
+class TimeTourViewBase(APIView):
+    permission_classes = [IsAuthenticated, OwnerAccount]
+    serializer_class = StudOnTourSerializer
+
+    async def finalize_response(self, request, response, *args, **kwargs):
+        if asyncio.iscoroutine(response):
+            # Wait for the coroutine to finish and return its result
+            response = await response
+            return super().finalize_response(request, response, *args, **kwargs)
+
+            # If the response is not a coroutine, return it as is
+        return super().finalize_response(request, response, *args, **kwargs)
+
+    @sync_to_async
+    def check_permissions(self, request):
+        super().check_permissions(request)
+
+    @sync_to_async
+    def check_object_permissions(self, request, obj):
+        super().check_object_permissions(request, obj)
+
+    @sync_to_async
+    def perform_authentication(self, request):
+        super().perform_authentication(request)
+
+    async def set_tour_time(self, request, student_on_tour_id, field_name, event_type):
+        student_on_tour_instance: StudentOnTour = await StudentOnTour.objects.filter(id=student_on_tour_id).afirst()
+        student: User = await sync_to_async(lambda: student_on_tour_instance.student)()
+        await self.check_object_permissions(request, student)
+
+        tz = pytz.timezone(config.settings.TIME_ZONE)
+        setattr(student_on_tour_instance, field_name, datetime.now(tz))
+
+        await student_on_tour_instance.asave()
+        channel_layer = get_channel_layer()
+        await channel_layer.group_send(
+            "student_on_tour_updates",
+            {
+                "type": event_type,
+                "student_on_tour_id": student_on_tour_instance.id,
+            },
+        )
+        return post_success(self.serializer_class(student_on_tour_instance))
+
+
+class StartTourView(TimeTourViewBase):
+    @extend_schema(responses=post_docs(TimeTourViewBase.serializer_class))
+    async def post(self, request, student_on_tour_id):
+        return await self.set_tour_time(request, student_on_tour_id, "started_tour", "student.on.tour.started")
+
+
+class EndTourView(TimeTourViewBase):
+    @extend_schema(responses=post_docs(TimeTourViewBase.serializer_class))
+    async def post(self, request, student_on_tour_id):
+        return await self.set_tour_time(request, student_on_tour_id, "completed_tour", "student.on.tour.completed")
+
+
+class ProgressTourView(APIView):
+    permission_classes = [IsAuthenticated, IsAdmin | IsSuperStudent]
+    serializer_class = ProgressTourSerializer
+
+    @extend_schema(responses=get_docs(serializer_class))
+    def get(self, request, student_on_tour_id):
+        student_on_tour = StudentOnTour.objects.get(id=student_on_tour_id)
+        serializer = ProgressTourSerializer(student_on_tour)
+        return Response(serializer.data)
```

Comment:
You could use `get_success()` from `request_response_util.py`

Reply:
You are totally right. This has been fixed now.